File: haskell/luhn/src/Luhn.hs
module Luhn (isValid) where

import Data.Char (ord, isDigit)

isValid :: String -> Bool
isValid n =
  let filteredDigit = filter isDigit n
      total = snd $ foldr parse (0, 0) filteredDigit
  in case filteredDigit of
       (x:[]) -> False
       otherwise -> total `mod` 10 == 0

parse :: Char -> (Int, Int) -> (Int, Int)
parse c (pos, total) =
  if pos `mod` 2 == 0 then
    (pos + 1, total + num)
  else
    (pos + 1, total + (double num))
  where double number = if number <= 4 then number * 2 else number * 2 - 9
        num = (ord c) - (ord '0')

File: python/luhn/luhn.py
class Luhn(object):
    def __init__(self, card_num):
        self.card_num = card_num
        self.valid = self.checksum(card_num)

    def checksum(self, card_num):
        sum_of_digits = 0
        pos = 0
        for num in card_num[::-1]:
            if num != ' ' and (num < '0' or num > '9'):
                return False
            if num >= '0' and num <= '9':
                n = int(num)
                if pos % 2 == 0:
                    sum_of_digits += n
                elif n <= 4:
                    sum_of_digits += n * 2
                else:
                    sum_of_digits += n * 2 - 9
                pos += 1
        return pos > 1 and sum_of_digits % 10 == 0

    def is_valid(self):
        return self.valid
